### PR TITLE
use getStatus to detect conflicted repository state

### PR DESCRIPTION
**Pull request by shiftkey**
## Overview

It's late, I can't sleep, and I'm restless from watching the news so I figure now is a great time to open a PR that changes how we detect merge conflicts in the app :tada:

This was started in #6082 but I wanted to move it out from there for more explicit discussion as it touches the core parts of conflict resolution flow.

## Description

Currently we look at the output of `git status` to see which files are conflicted, but if a user decides to stage files outside the app (whether they're resolved or not) the conflict status markers are forgotten by `git status`, so it's possible for Desktop to "forget" it's in the middle of a conflict.

This affects #6082 but also other scenarios that weren't the happy path flow, like the user addressing conflicts in the app but not completing the process by committing before switching back to Desktop.

Instead of relying on `git status` output, this PR switches us over to look for `.git/MERGE_HEAD` in the repository to identify when conflicts need to be resolved.

The flow of this information is then like this:

 - `getStatus` has a new field `mergeHeadFound` that indicates whether it found `MERGE_HEAD`
 - within `AppStore._loadStatus` we compute the new conflicted state for the repository (`getConflictStatus()`)
 - we update the repository state with the conflict information, as well as related stats ~~(still needs some cleanup)~~
 - inside `AppStore._triggerMergeConflictsFlow` we then check the current state of conflicts
 - we use this detail to display the merge conflicts dialog (if required)

This also removes some `_mergeConflictDetected()` which would set the conflicted state, but didn't use `git status` - this centralizes the responsibility to `AppStore._loadStatus()`

TODO:

 - [x] refactor the stats flow to reflect the desired behaviour
 - [x] a bunch more testing

## Release notes

Notes: no-notes

### CODE REVIEW

**Comment by billygriffin** [comment]
Cool! Couple questions:

1. Just clarifying: I assume based on the description that this is just relating to detecting when a conflict has already occurred, and therefore doesn’t help solve the mismatch between git merge and git merge-tree we’ve seen pop up occasionally in predicting conflicts.

2. Will this impact our metrics on conflicts resolved successfully vs. abandoned at all?

**Comment by shiftkey** [comment]
@billygriffin 

> 1. Just clarifying: I assume based on the description that this is just relating to detecting when a conflict has already occurred, and therefore doesn’t help solve the mismatch between git merge and git merge-tree we’ve seen pop up occasionally in predicting conflicts.

Yep, this is separate to the "merge hint" feature.

> 2. Will this impact our metrics on conflicts resolved successfully vs. abandoned at all?

I intend to keep that support around - based on the rules previously defined - and the code path gets hit whenever we update the status of the repository (like before) so there shouldn't be any change to that area.

**Comment by shiftkey** [comment]
I'd like to get reviewed in time for the next beta, to give it the maximum time to bake.

The PR that started this off #6082 won't be resolved without these changes, and I plan to open a fresh PR after this lands targeting the 1.5.0 release branch to ensure it gets in.

File: app/src/lib/stores/app-store.ts [hunk 1]
```diff
@@ -1609,7 +1546,51 @@ export class AppStore extends TypedBaseStore<IAppState> {
       return { workingDirectory, selectedFileIDs, diff }
     })
 
-    this._triggerMergeConflictsFlow(repository, status)
+    this.repositoryStateCache.updateChangesState(repository, state => {
```
Comment (outofambit) [inline app/src/lib/stores/app-store.ts]: can we pull this into its own method?

Comment (shiftkey) [reply app/src/lib/stores/app-store.ts]: Mind if I do this in a follow-up PR? There 's an existing [function above](https://github.com/desktop/desktop/blob/6d0206acc9f4ea543266acf948cf5064260f96e2/app/src/lib/stores/app-store.ts#L1488-L1546) that is also rather lengthy, and I'd love to tackle both of those at the same time.

**Comment by outofambit** [comment]
going to merge this in ~1hr so we can unblock #6122

**Comment by outofambit** [comment]
@shiftkey it looks like we have some conflicts now. would you mind resolving those? i'll merge this after that.

**Comment by shiftkey** [comment]
@outofambit done 🍎

**Pull request merged**